util: Type-safe transaction identifiers

dergoegge opened the pull request:
We currently have two different identifiers for transactions: `txid` (refering to the hash of a transaction without witness data) and `wtxid` (referring to the hash of a transaction including witness data). Both are typed as `uint256` which could lead to type-safety bugs in which one transaction identifier type is passed where the other would be expected.

This PR introduces explicit `Txid` and `Wtxid` types that (if used) would cause compilation errors for such type confusion bugs.

(Only the orphanage is converted to use these types in this PR)

dergoegge committed:
Introduce types for txids & wtxids

dergoegge committed:
Use type-safe txid types in orphanage

stickies-v reviewed (commented):
Concept ACK.

We don't consistently have clear variable naming (e.g. the ambiguous `hash`; or `txid` being used to refer to a wtxid), sometimes requiring quite a bit of developer time to figure out the expected ~type. I have experienced this quite a bit myself.

Introducing types makes this explicit at relatively low overhead, and I think the approach of subclassing `uint256` makes sense.

We'll have to be careful for this not to lead to unnecessary refactoring code churn, but that shouldn't be a reason not to make this change imo.

glozow commented:
I've definitely written/seen bugs that could be prevented if Txid and Wtxid were different types. When I review code it's something that I look for and it comes up surprisingly often. Interfaces taking `uint256` often implicitly expect them to be a txid, eg #23325.

Agree that checking at compile time would be nice and probably prevent bugs / make review easier. Concept ACK assuming we're ok with the amount of code change.

dergoegge commented:
> We'll have to be careful for this not to lead to unnecessary refactoring code churn, but that shouldn't be a reason not to make this change imo.

In terms of refactoring, this should be similar to the time type-safety PRs in that we can gradually convert our code.

dergoegge committed:
[net processing] Use HasWitness over comparing (w)txids

dergoegge commented:
Ready for review

luke-jr commented:
Concept ACK, but I'm not too sure about the `Wtxid` type. It's literally just a hash of the transaction, not really an identifier per se.

dergoegge commented:
> Concept ACK, but I'm not too sure about the Wtxid type.

Could you clarify your objections about the `Wtxid` type?  This PR aims to catch type confusion bugs (confusion between txids and wtxids) at compile-time and I don't see how to do that without these explicit types.

vasild commented on src/util/transaction_identifier.h:
```diff
@@ -0,0 +1,54 @@
+{
+    static constexpr bool negated_has_witness =
+        std::negation<std::bool_constant<has_witness>>::value;
+    using opposite_type =
+        transaction_identifier<negated_has_witness>;
+
+public:
+    transaction_identifier() : uint256{} {}
+    explicit transaction_identifier(const uint256& other) : uint256{other} {}
+    explicit transaction_identifier(const uint256&& other) : uint256{other} {}
+
+    // Delete copy/move constructors/assignment for differing transaction_identifier types.
+    transaction_identifier(const opposite_type&) = delete;
+    transaction_identifier(opposite_type&&) = delete;
+    transaction_identifier& operator=(const opposite_type&) = delete;
+    transaction_identifier& operator=(opposite_type&&) = delete;
+
+    // Disallow comparisons between different transaction_identifier types.
+    friend bool operator==(const transaction_identifier<has_witness>& a,
+                           const opposite_type& b) = delete;
+    friend bool operator!=(const transaction_identifier<has_witness>& a,
+                           const opposite_type& b) = delete;
+    friend bool operator<(const transaction_identifier<has_witness>& a,
+                          const opposite_type& b) = delete;
+    int Compare(const opposite_type&) const = delete;
+
+    // Allow comparisons between the same transaction_identifier type.
+    friend bool operator==(const transaction_identifier<has_witness>& a,
+                           const transaction_identifier<has_witness>& b) { return a.Compare(b) == 0; }
+    friend bool operator!=(const transaction_identifier<has_witness>& a,
+                           const transaction_identifier<has_witness>& b) { return a.Compare(b) != 0; }
+    friend bool operator<(const transaction_identifier<has_witness>& a,
+                          const transaction_identifier<has_witness>& b) { return a.Compare(b) < 0; }
+    constexpr int Compare(const transaction_identifier<has_witness>& other) const { return static_cast<const uint256&>(*this).Compare(other); }
+};
+
+/** Txid commits to all transaction fields except the witness. */
+using Txid = transaction_identifier<false>;
+/** Wtxid commits to all transaction fields including the witness. */
+using Wtxid = transaction_identifier<true>;
```
Is it not possible to replace all of that with just:

```cpp
class Txid : public uint256
{
};

class Wtxid : public uint256
{
};
```

It achieves the purpose of not allowing conversion and comparison between `Txid` and `Wtxid` and each one of them can be treated as `uint256`.

dergoegge replied:
That is what I originally had in this PR but it actually doesn't achieve the same on all fronts, so I decided to make it explicit to avoid subtle bugs.

e.g. the following was still permitted:
```c++
Txid txid;
Wtxid wtxid{txid};
// or
wtxid.Compare(txid);
// or
bool equal = (wtxid == txid); // same for !=
```

instagibbs reviewed (approved):
ACK 76fe2c304201646d63b1b01b397e18a99252a2d9

played around with some extensions to see where it could go, I think longer term we could use `GenTxid` for situations where both are acceptable, and `Wtxid/Txid` when only one type is, and slowly migrate towards that. f.e. `GenTxid` could have a `GetTypedHash` to convert to the other.

dergoegge commented:
@MarcoFalke could you give this another look? You've done similar work on our time type-safety, so I'd appreciate your feedback.

maflcko reviewed (commented):
left some quick comments

dergoegge commented on src/util/transaction_identifier.h:
```diff
@@ -0,0 +1,54 @@
+    static constexpr bool negated_has_witness =
+        std::negation<std::bool_constant<has_witness>>::value;
+    using opposite_type =
+        transaction_identifier<negated_has_witness>;
```
I'm OK with going either way but disallowing uint256 comparisons in this PR causes a lot of churn, i.e. going around casting all the uint256 to the right type. Most of these casts would later be reverted in follow up PRs that convert more things to use the new types (e.g. mempool, net processing, wallet).

I think I'd prefer doing the switch to the new type system first and then disallowing uint256 comparisons afterwards. What do you think?

dergoegge replied:
> The "opposite" type would be any type other than the type itself.

I meant opposite w.r.t `has_witness`, if you have a better name in mind, let me know.

> I just wanted to mention that this is not type safe and maybe other reviewers would want to know too?

I'll just add a todo comment about uint256 comparisons.

dergoegge replied:
Yea that works, applied in the latest push

dergoegge commented on src/util/transaction_identifier.h:
```diff
@@ -0,0 +1,54 @@
+    transaction_identifier(const opposite_type&) = delete;
+    transaction_identifier(opposite_type&&) = delete;
+    transaction_identifier& operator=(const opposite_type&) = delete;
+    transaction_identifier& operator=(opposite_type&&) = delete;
```
I kinda like making this explicit but the assignment operators are not necessary so i removed them

dergoegge commented on src/util/transaction_identifier.h:
```diff
@@ -0,0 +1,54 @@
+                           const transaction_identifier<has_witness>& b) { return a.Compare(b) != 0; }
+    friend bool operator<(const transaction_identifier<has_witness>& a,
+                          const transaction_identifier<has_witness>& b) { return a.Compare(b) < 0; }
+    constexpr int Compare(const transaction_identifier<has_witness>& other) const { return static_cast<const uint256&>(*this).Compare(other); }
```
Done

dergoegge commented on src/util/transaction_identifier.h:
```diff
@@ -0,0 +1,54 @@
+using Wtxid = transaction_identifier<true>;
+
+#endif // BITCOIN_UTIL_TRANSACTION_IDENTIFIER_H
+
```
Done

dergoegge commented on src/txorphanage.cpp:
```diff
@@ -20,7 +21,7 @@ bool TxOrphanage::AddTx(const CTransactionRef& tx, NodeId peer)
     LOCK(m_mutex);
 
-    const uint256& hash = tx->GetHash();
+    const auto& hash = tx->GetHash();
```
Done

dergoegge commented on src/util/transaction_identifier.h:
```diff
@@ -0,0 +1,53 @@
+
+    // Delete copy/move constructors/assignment for differing transaction_identifier types.
+    transaction_identifier(const opposite_type&) = delete;
+    transaction_identifier(opposite_type&&) = delete;
```
> Alternatively it could be disallowed properly.

converting from `uint256` can now only be done using `{Txid,Wtxid}::FromUint256`. This required less changes then expected so I think that is actually preferable.

dergoegge replied:
Added a comment to `transaction_identifier` to note that it is not fully type-safe yet due to allowing implicit conversions to uint256 in the interim.

glozow commented:
Rebase pls?

dergoegge commented:
Can someone restart the win64 job? failure is unrelated

ismaelsadeeq commented:
Concept ACK

BrandonOdiwuor commented:
Concept ACK

The introduction of transaction_identifier (including Txid and Wtxid) is a valuable enhancement that enforces compile-time checks. This proactive approach will significantly reduce the likelihood of runtime bugs caused by inadvertently confusing Txid and Wtxid. It also adds more clarity to the codebase.

BrandonOdiwuor reviewed (commented):
Concept ACK

The introduction of transaction_identifier (including Txid and Wtxid) is a valuable enhancement that enforces compile-time checks. This proactive approach will significantly reduce the likelihood of runtime bugs caused by inadvertently confusing Txid and Wtxid. It also adds more clarity to the codebase.

LarryRuane commented on src/util/transaction_identifier.h:
```diff
@@ -0,0 +1,61 @@
+public:
+    transaction_identifier() : uint256{} {}
+
+    template <typename Other>
+    transaction_identifier(const Other& other)
+        : uint256{other}
+    {
+        static_assert(std::is_same_v<Other, transaction_identifier<has_witness>>,
+                      "Forbidden copy type");
+    }
```
This method seems to be unused -- remove for now?

(Perhaps `FromUint256()` has taken its place?)

dergoegge replied:
I would like to keep this to make the allowed copy types explicit. This can be removed once we disallow `uint256 a = Wtxid/Txid{}`.

LarryRuane commented on src/util/transaction_identifier.h:
```diff
@@ -0,0 +1,61 @@
+
+    static transaction_identifier FromUint256(const uint256& uint)
+    {
+        auto id{reinterpret_cast<const transaction_identifier&>(uint)};
+        return id;
```
nit
```suggestion
        return reinterpret_cast<const transaction_identifier&>(uint);
```

mzumsande commented on src/net_processing.cpp:
```diff
@@ -5694,17 +5695,20 @@ bool PeerManagerImpl::SendMessages(CNode* pto)
-                        const uint256& hash = peer->m_wtxid_relay ? txinfo.tx->GetWitnessHash() : txinfo.tx->GetHash();
-                        CInv inv(peer->m_wtxid_relay ? MSG_WTX : MSG_TX, hash);
-                        tx_relay->m_tx_inventory_to_send.erase(hash);
+                        CInv inv{MSG_TX, txinfo.tx->GetHash()};
+                        if (peer->m_wtxid_relay) {
+                            inv = CInv{MSG_WTX, txinfo.tx->GetWitnessHash()};
+                        }
```
We need a different inv type `MSG_WTX` for wtxids!
Although the order could be switched (first create `MSG_WTX` and maybe overwrite that), since wtxid-based relay should be the standard by now and txid the exception.

dergoegge replied:
Switched the order

dergoegge commented:
> Rebase pls?

@glozow review pls?

stickies-v reviewed (commented):
Approach ACK c4c7b6ea9fe2e7ba89a994127073c54af30d6f88
_(did a fair amount of code review too so not expecting too many more comments there, looks great)_

I think the gradual rollout is definitely the way to go.

I've been trying to see if using a wrapped `uint256 m_hash` (without inheritance) makes more sense, but given that we want all `transaction_identifier` subclasses and `uin256` have an identical but non-interchangeable interface, I just ended up with more overhead, so that seems like a no-go.

Private inheritance seems the most natural approach in the long term, but seems like it is incompatible with a gradual rollout, as we can't use a conversion operator. 

So, I think it makes sense to stick with the public inheritance approach used now, and switch to private inheritance once all code is updated to the stricter types. This switch should be very limited in scope anyway, so I don't see an issue with doing things gradually.

dergoegge commented:
Changed the approach, so that `Wtxid` and `Txid` are now wrapped `uint256`s. A conversion function to `uint256` was added to allow for smoother transition to these types. Once we have converted most of our code we can simply delete the conversion function for full type-safety. @maflcko this is what you suggested quite early on, what do you think?

maflcko reviewed (approved):
lgtm

dergoegge commented on src/util/transaction_identifier.h:
```diff
@@ -0,0 +1,71 @@
+     * TODO: This should be removed once the majority of the code has switched
+     * to using the Txid and Wtxid types. Until then it makes for a smoother
+     * transition to allow this conversion. */
+    operator const uint256&() const { return m_wrapped; }
```
TIL

```c++
class foo {};

foo get_foo() { return foo{}; }

void main() {
    const foo& f = get_foo();
}
```

Assigning to a **const reference** extends the lifetime of a temporary object (in this case the result from `get_foo`) to the end of the function.

However, this is not the case when using conversion functions:

```c++
class foo {
    std::string m_str;
public:
    operator const std::string&() const { return m_str; }
};

foo get_foo() { return foo{}; }

void main() {
    const std::string& f = get_foo(); // this is a big no no
}
```

dergoegge replied:
It now returns a copy, which results in a few compiler errors/warnings but those were trivial to fix.

dergoegge replied:
> You can try (for fun) LIEFTIMEBOUND, to see if it catches this? 

It actually does, good stuff!

dergoegge commented on src/util/transaction_identifier.h:
```diff
@@ -0,0 +1,71 @@
+    static transaction_identifier FromUint256(const uint256& id) { return transaction_identifier{id}; }
+
+    /** Wrapped `uint256` methods. */
+    constexpr bool IsNull() { return m_wrapped.IsNull(); }
```
Done

glozow reviewed (commented):
Approach ACK, lgtm

dergoegge commented on src/util/transaction_identifier.h:
```diff
@@ -0,0 +1,71 @@
+     *
+     * Note: new code should use `ToUint256`.
+     *
+     * TODO: This should be removed once the majority of the code has switched
+     * to using the Txid and Wtxid types. Until then it makes for a smoother
```
> are we expecting to change everything to Txid or Wtxid?

I think almost everything using uint256 for transaction ids will be converted (including GenTxid which can just be a `std::variant<Txid, Wtxid>`) but there are probably a few exceptions.

I don't expect `m_recent_confirmed_transactions` or `m_recent_rejects` etc. to change as the underlying data structure (bloom filter) does not deal with txids, it simply accepts some bytes (`Span<const uint8_t>`). So converting the new types to a span (e.g. using `ToUint256`) should be fine.

instagibbs replied:
If a specific bloom filter is only meant to take one or the other I could potentially see them using the types more directly, but that would be a case by case basis clearly.

stickies-v commented on src/txorphanage.cpp:
```diff
@@ -159,7 +160,7 @@ void TxOrphanage::AddChildrenToWorkSet(const CTransaction& tx)
             for (const auto& elem : it_by_prev->second) {
                 // Get this source peer's work set, emplacing an empty set if it didn't exist
                 // (note: if this peer wasn't still connected, we would have removed the orphan tx already)
-                std::set<uint256>& orphan_work_set = m_peer_work_set.try_emplace(elem->second.fromPeer).first->second;
+                std::set<Txid>& orphan_work_set = m_peer_work_set.try_emplace(elem->second.fromPeer).first->second;
```
nit: That's a lot of `first`'s and `second`'s. This PR doesn't make it worse, but if you wanna sneak in some readability improvements I'd very happily support that, e.g.

<details>
<summary>git diff on 68b9c2f268e7a924f4c9182ab1b99c30af86f5d7</summary>

```diff
diff --git a/src/txorphanage.cpp b/src/txorphanage.cpp
index d4c07b0543..80ae4af662 100644
--- a/src/txorphanage.cpp
+++ b/src/txorphanage.cpp
@@ -158,13 +158,16 @@ void TxOrphanage::AddChildrenToWorkSet(const CTransaction& tx)
         const auto it_by_prev = m_outpoint_to_orphan_it.find(COutPoint(tx.GetHash(), i));
         if (it_by_prev != m_outpoint_to_orphan_it.end()) {
             for (const auto& elem : it_by_prev->second) {
+                const auto& txid = elem->first;
+                const auto& orphan_tx = elem->second;
                 // Get this source peer's work set, emplacing an empty set if it didn't exist
                 // (note: if this peer wasn't still connected, we would have removed the orphan tx already)
-                std::set<Txid>& orphan_work_set = m_peer_work_set.try_emplace(elem->second.fromPeer).first->second;
+                auto [it, inserted] = m_peer_work_set.try_emplace(orphan_tx.fromPeer);
+                std::set<Txid>& orphan_work_set = it->second;
                 // Add this tx to the work set
-                orphan_work_set.insert(elem->first);
+                orphan_work_set.insert(txid);
                 LogPrint(BCLog::TXPACKAGES, "added %s (wtxid=%s) to peer %d workset\n",
-                         tx.GetHash().ToString(), tx.GetWitnessHash().ToString(), elem->second.fromPeer);
+                         tx.GetHash().ToString(), tx.GetWitnessHash().ToString(), orphan_tx.fromPeer);
             }
         }
     }

```
</details>

glozow replied:
wdym readability. You look up the `elem` corresponding to the outpoint in this map, and `elem` refers to an iterator into another map, which you dereference to get the value of the `OrphanTx` object, so that you can extract the key to look up in a third map, so that you can obtain a reference to the element you just inserted or was already there.

dergoegge replied:
Not gonna touch that in this PR

stickies-v reviewed (approved):
ACK 68b9c2f268e7a924f4c9182ab1b99c30af86f5d7

The wrapped approach is less overhead than I thought it would be, and makes for a pretty simple implementation. Nice.

instagibbs reviewed (approved):
ACK 68b9c2f268e7a924f4c9182ab1b99c30af86f5d7

errors on compilation leave some readability to be desired, but what's new about C++

glozow reviewed (commented):
ACK 68b9c2f268e7a924f4c9182ab1b99c30af86f5d7, happy to reack if you update

hebasto commented:
Concept ACK.

dergoegge commented on src/util/transaction_identifier.h:
```diff
@@ -0,0 +1,71 @@
+    std::string ToString() const { return m_wrapped.ToString(); }
+    constexpr const unsigned char* data() const { return m_wrapped.data(); }
+    constexpr const unsigned char* begin() const { return m_wrapped.begin(); }
+    constexpr const unsigned char* end() const { return m_wrapped.end(); }
```
Ok great gonna include this

dergoegge commented on src/util/transaction_identifier.h:
```diff
@@ -0,0 +1,71 @@
+{
+    uint256 m_wrapped;
+
+    transaction_identifier(const uint256& wrapped) : m_wrapped{wrapped} {}
```
Done

dergoegge commented on src/validation.cpp:
```diff
@@ -616,7 +616,7 @@ class MemPoolAccept
         const CTransactionRef& m_ptx;
         /** Txid. */
-        const uint256& m_hash;
+        const Txid& m_hash;
```
The identifiers are now exported from `primitives/transaction`.

dergoegge commented on src/util/transaction_identifier.h:
```diff
@@ -0,0 +1,71 @@
+        } else {
+            static_assert(ALWAYS_FALSE<Other>, "Forbidden comparison type");
+        }
+    }
```
Done

dergoegge commented on src/net_processing.cpp:
```diff
@@ -5694,17 +5694,20 @@ bool PeerManagerImpl::SendMessages(CNode* pto)
                     LOCK(tx_relay->m_bloom_filter_mutex);
 
                     for (const auto& txinfo : vtxinfo) {
-                        const uint256& hash = peer->m_wtxid_relay ? txinfo.tx->GetWitnessHash() : txinfo.tx->GetHash();
```
Done

dergoegge commented:
Thank you all for the reviews! Addressed all your comments. This should be ready to go after re-acks.

hebasto commented on src/util/transaction_identifier.h:
```diff
@@ -0,0 +1,67 @@
+    bool operator<(const Other& other) const { return Compare(other) < 0; }
+
+    uint256 ToUint256() const { return m_wrapped; }
+    static transaction_identifier FromUint256(const uint256& id) { return transaction_identifier{id}; }
```
nit: If you'll touch this PR, the `transaction_identifier` might be dropped:
```suggestion
    static transaction_identifier FromUint256(const uint256& id) { return {id}; }
```

stickies-v commented on src/net_processing.cpp:
```diff
@@ -5694,17 +5694,19 @@ bool PeerManagerImpl::SendMessages(CNode* pto)
                         CInv inv(peer->m_wtxid_relay ? MSG_WTX : MSG_TX, hash);
                         tx_relay->m_tx_inventory_to_send.erase(hash);
+
+                        tx_relay->m_tx_inventory_to_send.erase(inv.hash);
```
Isn't this the same as `tx_relay->m_tx_inventory_to_send.erase(hash);` two lines up?

dergoegge replied:
whoops, removed

stickies-v commented on src/net_processing.cpp:
```diff
@@ -5694,17 +5694,19 @@ bool PeerManagerImpl::SendMessages(CNode* pto)
 
                     for (const auto& txinfo : vtxinfo) {
-                        const uint256& hash = peer->m_wtxid_relay ? txinfo.tx->GetWitnessHash() : txinfo.tx->GetHash();
+                        const uint256& hash = peer->m_wtxid_relay ? txinfo.tx->GetWitnessHash().ToUint256() : txinfo.tx->GetHash().ToUint256();
                         CInv inv(peer->m_wtxid_relay ? MSG_WTX : MSG_TX, hash);
```
nit: think this is a bit cleaner
```suggestion
                        CInv inv{
                            peer->m_wtxid_relay ? MSG_WTX                                 : MSG_TX
                            peer->m_wtxid_relay ? txinfo.tx->GetWitnessHash().ToUint256() : txinfo.tx->GetHash().ToUint256();
                        }
```

stickies-v commented on src/primitives/transaction.cpp:
```diff
@@ -12,6 +12,7 @@
 #include <tinyformat.h>
 #include <uint256.h>
 #include <util/strencodings.h>
+#include <util/transaction_identifier.h>
```
nit: no longer needed here. In `txorphanage.cpp` and `test/orphanage_tests.cpp`, `#include <util/transaction_identifier.h>` should be replaced with `#include <primitives/transaction.h>` since we need that include already anyway.

hebasto commented on src/util/transaction_identifier.h:
```diff
@@ -0,0 +1,67 @@
+    constexpr int Compare(const Other& other) const
+    {
+        static_assert(ALWAYS_FALSE<Other>, "Forbidden comparison type");
+        return 0;
```
I wonder should we use the `[[noreturn]]` attribute instead?

stickies-v reviewed (approved):
ACK 940a49978c70453e1aaf2c4a0bcb382872b844a5

hebasto commented:
A type safety implies that the following code:
```C++
CMutableTransaction tx;
uint256 hash = tx.GetHash();
```
must fail to compile. But it does.

If it is expected, maybe document such behaviour?

dergoegge commented:
> If it is expected, maybe document such behaviour?

It is documented here: https://github.com/bitcoin/bitcoin/blob/940a49978c70453e1aaf2c4a0bcb382872b844a5/src/util/transaction_identifier.h#L52-L59

hebasto reviewed (approved):
ACK 940a49978c70453e1aaf2c4a0bcb382872b844a5, I have reviewed the code and it looks OK.

~nit in 940a49978c70453e1aaf2c4a0bcb382872b844a5: using `auto` instead of explicit iterator types might make the code more concise, readable and maintainable.~

EDIT: I saw https://github.com/bitcoin/bitcoin/pull/28107#discussion_r1355379212

instagibbs reviewed (approved):
re-ACK 940a49978c70453e1aaf2c4a0bcb382872b844a5

BrandonOdiwuor reviewed (commented):
re-ACK 940a49978c70453e1aaf2c4a0bcb382872b844a5

glozow reviewed (commented):
reACK 940a49978c
range-diff from last review
- export Txid and Wtxid https://github.com/bitcoin/bitcoin/pull/28107#discussion_r1355310789
- no reassignment of inv https://github.com/bitcoin/bitcoin/pull/28107#discussion_r1355293074
- comment telling people to use `FromUint256` https://github.com/bitcoin/bitcoin/pull/28107#discussion_r1355415958
- overloaded instead of templated `Compare` https://github.com/bitcoin/bitcoin/pull/28107#discussion_r1355126839
- removing unnecessary autoing https://github.com/bitcoin/bitcoin/pull/28107#discussion_r1355379212
- includes
- dropping unnecessary https://github.com/bitcoin/bitcoin/pull/28107#discussion_r1356624512
- using `std::byte` instead of `unsigned char` + needed casts https://github.com/bitcoin/bitcoin/pull/28107#discussion_r1350254408

glozow commented:
Assuming we should wait until branch off to merge?

fanquake commented:
> Assuming we should wait until branch off to merge?

Any particular reason? If these start getting used more widely, it's just going to make doing backports more complicated if so.

hebasto commented on src/txorphanage.cpp:
```diff
@@ -100,10 +101,10 @@ void TxOrphanage::EraseForPeer(NodeId peer)
 
     int nErased = 0;
-    std::map<uint256, OrphanTx>::iterator iter = m_orphans.begin();
+    auto iter = m_orphans.begin();
```
My opinion is the [opposite](https://github.com/bitcoin/bitcoin/pull/28107#pullrequestreview-1674099958).

However, I agree with the current code.

glozow commented:
> > Assuming we should wait until branch off to merge?
> 
> Any particular reason? If these start getting used more widely, it's just going to make doing backports more complicated if so.

Idc that much, was asked why I didn't merge yet. Conversions are still possible so it shouldn't be that complicated, but whatever you prefer. Could use this time to look at what it looks like to change the rest of the codebase and see if we run into any bumps.

ajtowns commented on src/util/transaction_identifier.h:
```diff
@@ -0,0 +1,67 @@
+     * to using the Txid and Wtxid types. Until then it makes for a smoother
+     * transition to allow this conversion. */
+    operator uint256() const { return m_wrapped; }
+};
```
Worth adding a `bool HasWitness() const { return has_witness; }` ?

dergoegge replied:
I don't see when this would be needed, do you have a use case?

ajtowns replied:
Just seems like good practice to expose the information. Can always be added later when needed, I suppose.

dergoegge commented on src/util/transaction_identifier.h:
```diff
@@ -0,0 +1,67 @@
+    uint256 m_wrapped;
+
+    // Note: Use FromUint256 externally instead.
+    transaction_identifier(const uint256& wrapped) : m_wrapped{wrapped} {}
```
That might be better but would allow something like `Txid{Wtxid{}}` right now (because of the conversion function). We could make it a template like we do for `Compare`...

ajtowns replied:
But (a) that would presumably be silly to write, and (b) the conversion function is going away, which will ensure any code like that also goes away.

dergoegge commented on src/util/transaction_identifier.h:
```diff
@@ -0,0 +1,67 @@
+     * TODO: This should be removed once the majority of the code has switched
+     * to using the Txid and Wtxid types. Until then it makes for a smoother
+     * transition to allow this conversion. */
+    operator uint256() const { return m_wrapped; }
```
See https://github.com/bitcoin/bitcoin/pull/28107#discussion_r1350398367

I'm not sure if handing out a reference to the underlying uint256 is a great idea, i guess it could work with `LIFETIMEBOUND`.
Copying seems safer and shouldn't happen all that often once we are using the types in most places.

ajtowns replied:
Well, this is going away anyway, so sure. Feels to me like `foo& x = create_a_new_temporary();` is already a weird thing to do though.

dergoegge commented on src/util/transaction_identifier.h:
```diff
@@ -0,0 +1,67 @@
+#ifndef BITCOIN_UTIL_TRANSACTION_IDENTIFIER_H
+#define BITCOIN_UTIL_TRANSACTION_IDENTIFIER_H
```
No strong opinion, what do the other reviewers think?

dergoegge commented on src/net_processing.cpp:
```diff
@@ -1852,7 +1852,7 @@ void PeerManagerImpl::BlockConnected(const std::shared_ptr<const CBlock>& pblock
         for (const auto& ptx : pblock->vtx) {
             m_recent_confirmed_transactions.insert(ptx->GetHash());
-            if (ptx->GetHash() != ptx->GetWitnessHash()) {
+            if (ptx->HasWitness()) {
```
Comparing `Txid` and `Wtxid` is no longer allowed. We could use `ToUint256` or add a cache for `HasWitness`?

I like using `HasWitness` because it is immediately obvious what we are checking.

glozow replied:
Can `CTransaction::HasWitness` be changed to check whether `hash != m_witness_hash`?

ajtowns replied:
Could move the "iterate through" part into `ComputeWitnessHash`, and have `HasWitness` just return `hash.ToUint256() == m_witness_hash.ToUint256()`. Note that that's an example of where `ToUint256()` not doing a copy would make sense.

dergoegge commented:
I'll open a follow up PR for @ajtowns's suggestions soon.
